lib/tensorflow/tensor.rb

```diff
@@ -182,17 +182,39 @@ def ruby_array_to_c(array, type)
   #   - Value of the element contained in the specified position in the tensor.
   #
   def getval(dimension)
-    raise("Invalid dimension array passed as input.",ShapeError) if dimension.length != dimensions.length
+    raise("Invalid dimension array passed as input.",ShapeError) if dimension.length != shape.length
     (0..dimension.length-1).each do |i|
-      raise("Invalid dimension array passed as input.",ShapeError) if dimension[i] > dimensions[i] || dimension[i] < 1 || !(dimension[i].is_a? Integer)
+      raise("Invalid dimension array passed as input.",ShapeError) if dimension[i] > shape[i] || dimension[i] < 1 || !(dimension[i].is_a? Integer)
     end
-    sum = dimension[dimension.length - 1]  - 1
-    prod = dimensions[dimensions.length - 1]
+    sum = dimension.last - 1
+    prod = shape.last
     (0..dimension.length - 2).each do |i|
        sum += (dimension[dimension.length - 2 - i] - 1) * prod
-       prod *= dimensions[dimensions.length - 2 - i]
+       prod *= shape[shape.length - 2 - i]
     end
 
     flatten[sum]
   end
+
+  private
+
+  #
+  # Recursively finds the shape of the input array.
+  #
+  # * *Returns* :
+  #   - Dimension array `[[2], [4]].shape` => `[2, 1]`
+  #
+  # TODO: Handle non-rectangular arrays and raise error if mixed data types
+  def self.set_shape(value)
```

Comment:
Hi @zverok, thanks I'm familiar with the Ruby conventions and I guess it could also be called `shape_of()` instead. There's already a bare `.shape`-method (the one renamed from `.dimensions`) which is the getter for the tensor-shape. So the reason to prefix `set_`, is that if you look in the [Python-implementation](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/framework/ops.py#L378) the shape is inferred and _set_ when the tensor is initialized, or when someone later explicitly calls `set_shape`, and the result is held in a private variable. My idea was to follow the Tensorflow team's own design implementation and move [Line 45](https://github.com/somaticio/tensorflow.rb/blob/master/lib/tensorflow/tensor.rb#L45)

```ruby
# lib/tensorflow/tensor.rb#L45
self.shape = self.class.set_shape(value)
```
into the `set_shape` method. Thereby the shape would be _set_ only when this method was called during initialization or when someone _later_ wanted to overwrite the shape value.